repose-aggregator/components/filters/openapi-validator-filter/src/main/scala/org/openrepose/filters/openapivalidator/OpenApiValidatorFilter.scala

```diff
@@ -0,0 +1,205 @@
+    "validation.request.security.invalid" -> ValidationIssue(HttpServletResponse.SC_BAD_REQUEST, 3),
+    "validation.request.parameter.header.missing" -> ValidationIssue(HttpServletResponse.SC_BAD_REQUEST, 4),
+    "validation.request.parameter.query.missing" -> ValidationIssue(HttpServletResponse.SC_BAD_REQUEST, 5),
+    "validation.request.parameter.missing" -> ValidationIssue(HttpServletResponse.SC_BAD_REQUEST, 6),
+    "validation.request.parameter.enum.invalid" -> ValidationIssue(HttpServletResponse.SC_BAD_REQUEST, 7),
+    "validation.request.parameter.collection.invalid" -> ValidationIssue(HttpServletResponse.SC_BAD_REQUEST, 8),
+    "validation.request.parameter.collection.invalidFormat" -> ValidationIssue(HttpServletResponse.SC_BAD_REQUEST, 9),
+    "validation.request.parameter.collection.tooManyItems" -> ValidationIssue(HttpServletResponse.SC_BAD_REQUEST, 10),
+    "validation.request.parameter.collection.tooFewItems" -> ValidationIssue(HttpServletResponse.SC_BAD_REQUEST, 11),
+    "validation.request.parameter.collection.duplicateItems" -> ValidationIssue(HttpServletResponse.SC_BAD_REQUEST, 12),
+    "validation.request.contentType.invalid" -> ValidationIssue(HttpServletResponse.SC_BAD_REQUEST, 13),
+    "validation.request.contentType.notAllowed" -> ValidationIssue(HttpServletResponse.SC_UNSUPPORTED_MEDIA_TYPE, 14),
+    "validation.request.accept.invalid" -> ValidationIssue(HttpServletResponse.SC_BAD_REQUEST, 15),
+    "validation.request.accept.notAllowed" -> ValidationIssue(HttpServletResponse.SC_NOT_ACCEPTABLE, 16),
+    "validation.schema.invalidJson" -> ValidationIssue(HttpServletResponse.SC_BAD_REQUEST, 17),
+    "validation.schema.unknownError" -> ValidationIssue(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, 18)
+  )
+
+  case class ValidationIssue(statusCode: Int, priority: Int)
+
+  case class ValidationFailure(issue: ValidationIssue, message: ValidationReport.Message)
+
+  /**
+    * @param httpServletRequest an [[HttpServletRequest]]
+    * @return an [[HttpServletRequest]] with a [[ServletInputStream]] that supports
+    *         [[java.io.InputStream#mark]] and [[java.io.InputStream#reset]]
+    */
+  private def getBufferingHttpServletRequest(httpServletRequest: HttpServletRequest): HttpServletRequest = {
+    val servletInputStream = httpServletRequest.getInputStream
+    if (servletInputStream.markSupported) {
+      httpServletRequest
+    } else {
+      new HttpServletRequestWrapper(httpServletRequest, new BufferedServletInputStream(servletInputStream))
+    }
+  }
+
+  /**
+    * Maps all Error-level Messages to Issues, sorts the Issues by priority, and returns the highest priority Issue,
+    * if one is present.
+    *
+    * Note that a report may contain Messages for more than one failed check.
+    *
+    * Note that we have chosen to ignore unknown issues (i.e., remove issues that are not present in our issue map).
```

Comment:
It does add that burden, but that was intentional. Fortunately, the burden to add support on our end is fairly minimal. 

If we were to address (rather than drop) unknown issues, we lose some stability; if, for example, we want to upgrade the library to take advantage of a performance improvement, we may end up pulling in new issues that we don't explicitly know about which may cause new failures that operators were not expecting. 

To give a bit more context, I actually made this decision while writing the filter. Before I had completed the mapping between servlet requests and validation requests, I would see validation issues caused by the incomplete mapping. By dropping those issues until support was added, I could verify more behavior. So put generally, if there is an issue in the request wrapper that would lead to false-positives on new, untested validation issues, I would rather not fail since it is not necessarily the case that the request itself is invalid. For example, some validation could be added for a specific header (say, `X-Encoded`), but our request wrapper class does not decode that header even though the validation library expects it to be decoded. Well, the value of `X-Encoded` may be totally valid, but we would still return an error response to the user, which does not seem right. Granted, that was a pretty weak example, but generally speaking, and mismatch of expectations between our code and the validation library when it comes to a validation would lead to that result.

So I guess generally speaking, I like how the decision to drop unknown issues decouples us a bit from the validation library. I also like that it forces us to be explicit about what we support since that means we will write tests for the things we support and can document those things.

Reply:
I understood the reasoning, I just didn't agree with it. we've explicitly said we are using this other library for validation, that means people are going to look at that library and what it supports and when we don't behave that way they are going to come to us asking why we say we support this thing but don't.
So my main problem comes to, the user has given us a document to use for validation, the validator is saying no this doesn't match what the user said is valid, and then we are saying looks good to me and passing it on. To me when dealing with validation the default should be to nope on out and not allow things through.
We should poke @wdschei and @Mario-Lopez to hear their thoughts.

Reply:
> I understood the reasoning, I just didn't agree with it. we've explicitly said we are using this other library for validation, that means people are going to look at that library and what it supports and when we don't behave that way they are going to come to us asking why we say we support this thing but don't.
> So my main problem comes to, the user has given us a document to use for validation, the validator is saying no this doesn't match what the user said is valid, and then we are saying looks good to me and passing it on. To me when dealing with validation the default should be to nope on out and not allow things through.
> We should poke @wdschei and @Mario-Lopez to hear their thoughts.

Fair enough.

To comment further on your main problem, it seems to me like we are already in a situation where the user has given us a document that asserts something that we cannot validate. For example, [multi-part form data validation is not currently supported by the validation library](https://bitbucket.org/atlassian/swagger-request-validator/src/b295b1aad09dbb214107340a7acc2a978db50813/swagger-request-validator-core/src/main/java/com/atlassian/oai/validator/interaction/request/RequestBodyValidator.java#lines-116:117). [Here is a full list of what is supported by the library we are using](https://bitbucket.org/atlassian/swagger-request-validator/src/b295b1aad09dbb214107340a7acc2a978db50813/docs/OPENAPIv3.md). Besides that, I am dubious of the claim that our users will not only read our documentation, but will dig into the documentation for the underlying library. And honestly, I would rather not make them.

In my view, our use of the `swagger-request-validator` library is an implementation detail, so it is left to us to decide what parts of that library we want to expose and how. In fact, upon reflection, I am tempted to remove the note in the docs for this filter about using that library. Instead, I would add a note explicitly stating that the issues I have already listed are the only validations we support.

From a code perspective, it is trivial to implement this either way -- whether we fail or pass on unknown issues.

Reply:
The team has decided to fail on unknown validation issues rather than pass. I will update this code to reflect that decision.